nvflare/app_common/dex/dxo_exchanger.py

```diff
@@ -0,0 +1,115 @@
+            heartbeat_timeout (float): how long to wait for a heartbeat from the peer before treating the peer as gone,
+                0 means DO NOT check for heartbeat.
+        """
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self._get_poll_interval = get_poll_interval
+
+        self.data_exchange_path = None
+        self._pipe_role = pipe_role
+        self._pipe_name = pipe_name
+        self._heartbeat_interval = heartbeat_interval
+        self._heartbeat_timeout = heartbeat_timeout
+        self.pipe_monitor = None
+        self._file_accessor_name = None
+
+    def initialize(self, data_exchange_path: str, file_accessor: Optional[FileAccessor] = None):
+        self.data_exchange_path = os.path.abspath(data_exchange_path)
+        file_pipe = FilePipe(self.data_exchange_path)
+        if file_accessor is not None:
+            file_pipe.set_file_accessor(file_accessor)
+        else:
+            file_pipe.set_file_accessor(PickleFileAccessor())
+
+        # use file accessor name to determine serialize/deserialize
+        self._file_accessor_name = file_pipe.accessor.__class__.__name__
+        file_pipe.open(self._pipe_name, self._pipe_role)
+        self.pipe_monitor = PipeMonitor(
+            file_pipe,
+            read_interval=self._get_poll_interval,
+            heartbeat_interval=self._heartbeat_interval,
+            heartbeat_timeout=self._heartbeat_timeout,
+        )
+        self.pipe_monitor.start()
+
+    def put(self, data_id: str, data: DXO):
+        if self.pipe_monitor is None:
+            raise RuntimeError("PipeMonitor is not initialized.")
+        req = Message.new_request(self._file_accessor_name, data, data_id)
+        self.pipe_monitor.send_to_peer(req)
+
+    def _get_file(self, req_topic: str, data_id: Optional[str] = None, timeout: Optional[float] = None):
+        if self.pipe_monitor is None:
+            raise RuntimeError("PipeMonitor is not initialized.")
+        start = time.time()
+        while True:
+            msg: Message = self.pipe_monitor.get_next()
+            if not msg:
+                if timeout and time.time() - start > timeout:
+                    self.pipe_monitor.notify_abort()
+                    raise TimeoutError(f"get file timeout after {timeout} secs")
+            elif msg.topic == Topic.HEARTBEAT:
+                continue
+            elif msg.topic == Topic.ABORT:
+                raise RuntimeError("the other end is aborted")
+            elif msg.topic in [Topic.END, Topic.PEER_GONE]:
+                raise RuntimeError(f"received {msg.topic}: {msg.data} while waiting for result for {req_topic}")
+            elif msg.topic == req_topic and data_id is None:
+                return msg.data
+            elif msg.topic == req_topic and msg.msg_id == data_id:
+                return msg.data
+            time.sleep(self._get_poll_interval)
+
+    def get(self, data_id: Optional[str] = None, timeout: Optional[float] = None):
```

Comment:
This data_id seems to be topic. It that's the case, then just use the term "topic", instead of "data_id".

Reply:
thanks, changed.